src/devices/src/virtio/net/tap.rs

```diff
@@ -89,9 +100,8 @@ impl IfReqBuilder {
 
     pub(crate) fn execute<F: AsRawFd>(mut self, socket: &F, ioctl: u64) -> Result<ifreq> {
         // SAFETY: ioctl is safe. Called with a valid socket fd, and we check the return.
-        let ret = unsafe { ioctl_with_mut_ref(socket, ioctl, &mut self.0) };
-        if ret < 0 {
-            return Err(Error::IoctlError(IoError::last_os_error()));
+        if unsafe { ioctl_with_mut_ref(socket, ioctl, &mut self.0) } < 0 {
+            return Err(Error::IfreqExecuteError(IoError::last_os_error()));
```

Comment:
I think there is space for improvement here. The TUNSETIFF ioctl returns EINVAL when a multiqueue tap backend has been configured and it also can return other various error types like (operation not permitted, bad file descriptor etc). So having a single message for all (i.e "check that your network device is configured properly") might actually be misleading. 

We could filter the type of error that the IOCTL returnes:
```
let err = IoError::last_os_error();
if err.kind() == std::io::ErrorKind::InvalidInput {
    return Error(IfReqInvalidArgument(err));
}
```
and the description of `IfReqInvalidArgument` could be something even more explicit, like:
`Invalid TUN/TAP Backend provided. Firecracker supports only TUN/TAP network backend with no multi queue support`.

Reply:
> Well, I think that if `TUNSETIFF` fails then there is definitely an issue with the configuration of the network device, so trying to catch only this specific one is incorrect. 

I think you misunderstood what I was proposing. I am not suggesting to catch only `InvalidInput`. I  was saying that in case of `InvalidInput` to explicitly let the user know what type of backend they should be configuring. 

> `InvalidInput` can also be caused by something else (even though in our case it is unlikely). Because of that I think having more generic error is better. It would be even better if there was some section in the documentation that states all possible net device configurations and I could add link to that section in the error message.

The initial issue we are trying to solve is to throw a proper error message when user configures a multi-queue backend, right? So, "having a more generic error" goes in the opposite direction IMO. I understand that a proper check would require too much effort for the gains and that is why I think that the least we can do is try to guide the user to a proper configuration of the backend. 

Your proposal to guide the user to our documentation is also a valid option IMO.
(Although, putting  a  link is not good practice since links are ephemerals -> locations change).
We can just say `Invalid TUN/TAP Backend provided. Check our documentation on setting up the network device (i.e network-setup).`
However, we do not mention in our net setup we do not support multi-queue. Would make sense for this issue to also be explicit about that there too, so that we avoid future inquiries in that direction.

Reply:
@dianpopa  By saying `more generic error is better` I basically meant that the error should look like your new proposal: 
`Invalid TUN/TAP Backend provided. Check our documentation on setting up the network device (i.e network-setup)` 
rather then:
`Invalid TUN/TAP Backend provided. Firecracker supports only TUN/TAP network backend with no multi queue support`

@mattschlebusch I think the problem is not the `IoError::last_os_error()` itself, but the lack of the proper description around this error. 
Also I don't think changing `jailer` for this issue is a great idea. Who knows what security nuances it would bring.

So it seems we all agree that guiding user to the documentation is a reasonable solution here.
I think the phrase: 
`Invalid TUN/TAP Backend provided. Check our documentation on setting up the network device (i.e network-setup)`
 that @dianpopa proposed is fine.
Now the question is where to put the documentation and how to reference it in the error message.

Reply:
We already have documentation on network setup (is found in docs/nework-setup.md).
But putting the location inside the error message is risky since that might change in the future. So, i think it suffices to leave the error message just referring to the doc.